File: starter_kits/C++/hlt/log.cpp
#include "log.hpp"

#include <fstream>
#include <iostream>
#include <vector>
#include <chrono>

static std::ofstream log_file;
static std::vector<std::string> log_buffer;
static bool has_opened = false;
static bool has_atexit = false;

void dump_buffer_at_exit() {
    if (has_opened) {
        return;
    }

    auto now_in_nanos = std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::high_resolution_clock::now().time_since_epoch()).count();
    std::string filename = "bot-unknown-" + std::to_string(now_in_nanos) + ".log";
    std::ofstream file(filename, std::ios::trunc | std::ios::out);
    for (const std::string& message : log_buffer) {
        file << message << std::endl;
    }
}

void hlt::log::open(int bot_id) {
    if (has_opened) {
        hlt::log::log("Error: log: tried to open(" + std::to_string(bot_id) + ") but we have already opened before.");
        exit(1);
    }

    has_opened = true;
    std::string filename = "bot-" + std::to_string(bot_id) + ".log";
    log_file.open(filename, std::ios::trunc | std::ios::out);

    for (const std::string& message : log_buffer) {
        log_file << message << std::endl;
    }
    log_buffer.clear();
}

void hlt::log::log(const std::string& message) {
    if (has_opened) {
        log_file << message << std::endl;
    } else {
        if (!has_atexit) {
            has_atexit = true;
            atexit(dump_buffer_at_exit);
        }
        log_buffer.push_back(message);
    }
}

File: starter_kits/Scala/hlt/Log.scala
package hlt

import java.io.{FileWriter, IOException}

import scala.collection.mutable

object Log {
  private val _buffer = new mutable.Queue[String]
  private var _logFileWriter: FileWriter = null

  def open(botId: Int) = {
    if (_logFileWriter != null) {
      Log.log(s"Error: log: tried to open($botId) but we have already opened before.")
      throw new IllegalStateException
    }
    val filename = s"bot-$botId.log"
    try {
      _logFileWriter = new FileWriter(filename)
      _buffer.foreach(line => _logFileWriter.append(line).append('\n'))
      _buffer.clear()
    } catch {
      case e: IOException =>
        throw new IllegalStateException(e)
    }
  }

  def log(message: String) = {
    if (_logFileWriter == null) {
      _buffer.enqueue(message)
    }
    else {
      try _logFileWriter.append(message).append('\n').flush()
      catch {
        case e: IOException =>
          e.printStackTrace()
      }
    }
  }

  private class AtExit extends Thread {
    override def run(): Unit = {
      if (_logFileWriter != null) return
      val now_in_nanos = System.nanoTime
      val filename = s"bot-unknown-$now_in_nanos.log"
      var writer: FileWriter = null;
      try {
        writer = new FileWriter(filename)
        _buffer.foreach(line => _logFileWriter.append(line).append('\n'))
      } catch {
        case e: IOException =>
        // Nothing much we can do here.
      } finally if (writer != null) writer.close()
    }
  }

  Runtime.getRuntime.addShutdownHook(new Log.AtExit)
}
